fgpyo/sam/__init__.py

```diff
@@ -1231,3 +1263,234 @@ class SamOrder(enum.Enum):
+        if (
+            rec.is_unmapped
+            or rec.cigarstring is None
+            or rec.query_sequence is None
+            or rec.query_qualities is None
+        ):
+            return
+
+        for hit in cls.many_from_rec(rec):
+            # TODO: When the original record has hard clips we must set the bases and quals to "*".
+            #       It would be smarter to pad/clip the sequence to be compatible with new cigar...
+            if "H" in rec.cigarstring:
+                query_sequence = NO_QUERY_BASES
+                query_qualities = None
+            elif rec.is_forward and not hit.is_forward:
+                query_sequence = reverse_complement(rec.query_sequence)
+                query_qualities = rec.query_qualities[::-1]
+            else:
+                query_sequence = rec.query_sequence
+                query_qualities = rec.query_qualities
+
+            secondary = AlignedSegment(header=rec.header)
+            secondary.query_name = rec.query_name
+            secondary.reference_id = rec.header.get_tid(hit.reference_name)
+            secondary.reference_name = hit.reference_name
+            secondary.reference_start = hit.reference_start
+            secondary.mapping_quality = 0 if hit.mapq is None else hit.mapq
+            secondary.cigarstring = str(hit.cigar)
+            secondary.query_sequence = query_sequence
+            secondary.query_qualities = query_qualities
+            secondary.is_read1 = rec.is_read1
+            secondary.is_read2 = rec.is_read2
+            secondary.is_duplicate = rec.is_duplicate
+            secondary.is_paired = rec.is_paired
+            secondary.is_proper_pair = False
+            secondary.is_qcfail = rec.is_qcfail
+            secondary.is_forward = hit.is_forward
+            secondary.is_secondary = True
+            secondary.is_supplementary = False
+            secondary.is_mapped = True
+
+            # NB: mate information on a secondary alignment points to mate/next primary alignment.
+            secondary.next_reference_id = rec.next_reference_id
+            secondary.next_reference_name = rec.next_reference_name
+            secondary.next_reference_start = rec.next_reference_start
+            secondary.mate_is_mapped = rec.mate_is_mapped
+            secondary.mate_is_reverse = rec.mate_is_reverse
+            secondary.set_tag("MC", rec.get_tag("MC") if rec.has_tag("MC") else None)
+            secondary.set_tag("MQ", rec.get_tag("MQ") if rec.has_tag("MQ") else None)
+            secondary.set_tag("ms", rec.get_tag("ms") if rec.has_tag("ms") else None)
+
+            # NB: set some optional but highly recommended SAM tags on the secondary alignment.
+            secondary.set_tag("AS", hit.alignment_score)
+            secondary.set_tag("NM", hit.edit_distance)
+            secondary.set_tag("RG", rec.get_tag("RG") if rec.has_tag("RG") else None)
+            secondary.set_tag("RX", rec.get_tag("RX") if rec.has_tag("RX") else None)
```

Comment:
I wish pysam allowed us to provide a default value (including `None`) to the `get_tag` method.  Perhaps an issue?